### ISSUE TITLE
Default resource listing for a bootstrapped instance

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#272]
This came up in the context of the in-job tool, but I'm thinking it would be worth having for the general case of bootstrapped instances.  Specifically an RDL, or RDL-equivalent, should be populated when the instance is created from the nodes/cores available in its execution environment. For our current targets, the node list would come from the SLURM environment, or it could come from PMI if that makes more sense, and the core listing can come from hwloc.

This way a user would not have to supply an RDL to use flux in a bootstrapped environment.

**Comment by grondo** [comment]
Early on we had the idea to have a kind of "distributed hwloc" inside each instance. Really passing around RDL subsets replaced this idea a bit (RDL itself would be a superset of the hwloc information), but you are totally right for SLURM-bootstrapped instance it makes much more sense to just autogenerate the RDL.

My inclination would be that SLURM_JOB_NODELIST isn't necessary, as each rank could gather its socket/memory/core information with hwloc and store this in kvs by rank. Sched could either load this into RDL using a shim, then load rdl as in normal case, or we could write a special case loader for the in-kvs values.

Also we'd need to ensure initial program doesn't start until resource gathering is complete.
I think there is an existing hook we could use, or we could explore the idea of initial program as init script and call the resource loader as a service from the init script, before calling user command(s)

**Comment by dongahn** [comment]
Absolutely - RDL will be a superset (e.g.,  other hierarchies like IO switch hierarchy etc.)

FYI -- As part of #221, I essentially created a named_reduction (#266)  although I isolated it to wrexec for now. This can merge the inputs from the specified number of nodes in a scalable way and store the merged output to the kvs at the root (the name used in the named reduction is the path in the kvs.)

If a scalable way to merge socket/memory/core to create a RDL in the boostrapped environment, this code can be leveraged as well.

**Comment by garlick** [comment]
+1 for initial program as init script.

Seems like we have a very basic start on this with `src/modules/wreck/resrc.c`.
```
 ./flux kvs dir -r resrc
resrc.rank.0.cores = 12
resrc.rank.0.mem = 24022
resrc.rank.0.alloc.cores = 0
```

**Comment by trws** [comment]
Since this is meant to be for a tool, I could probably build on that and make something that will generate an RDL from the resrc data if one isn't specified by the user.

The one trick here is that the RDL material is off in flux-sched, and I'm increasingly wanting to have access to at least a version of that in the core for a number of reasons.  I can see why the experimental scheduler components, simulator etc. are out there, but do we want to keep the base scheduler and RDL separate long-term?

**Comment by grondo** [comment]
@trws that is a good question and I don't really know the anser. I think we set it up this way because there were already starting to be multiple "RDL" implementations. Now there is an RDL that is tightly wound to a scheduler implementation so it makes sense to keep that version with the scheduler that uses it.

The Lua RDL prototype needs to go away or be stripped down to just a config parser if that is useful.

It feels like for the tool you are building a simpler, auto-generated rdl that is an expansion of the `resrc` implementation is the right direction. Maybe since this is already in flux-core, we can keep it that way?

What I'd like to see is a slight expansion of resrc (and maybe go ahead and buy some vowels and call it `resource`) that contains socket/numa information as well as the current information. If we settle on a "single-writer" model, the information of available/unavailable resources could be kept up to date in the kvs and query tools could easily be built against the kvs (i.e. in a just a few lines of python/Lua)

To get natural kvs-based reduction, maybe each rank's resource description could be a JSON blob, which is gauranteed to match for identical systems (maybe just a json transcription of lstopo or something) If all ranks write their blob then do a `kvs_fence` we might have a pretty scalable resource gather operation.

**Comment by lipari** [comment]
Having resources as part of the core makes sense to me.  As we specify and optionally confine a program to a set of resources, one expects the representation of those resources will be available in the core.  This would support immediate execution as well as well as the flux-sched project.

There is a resource-related construct does that not need to be in the core, namely the associated graphs (I'm not sure about the physical graph).  The graphs are used to schedule resources, but if one is going to specify resources under immediate execution, then the connectivity info the graphs provide is irrelevant.

**Comment by trws** [comment]
The graphs could be useful for a number of other things, not the least of which being reporting or correlation between resources.  Either way, I think @grondo's idea makes a lot of sense.  One thing I'm not sure how to handle though is where a user runs multiple brokers on a node.  We probably don't want each of them to use all cores independently.

**Comment by grondo** [comment]
@trws -- yeah, the main case here of course is `flux-start --size...` I
didn't have any good ideas there...

**Comment by grondo** [comment]
Though `flux-start` as the "bootstrap instance" could perhaps do something
like set cpu_affinity mask, which `resource` service could honor?

**Comment by garlick** [comment]
> To get natural kvs-based reduction, maybe each rank's resource description could be a JSON blob, which is gauranteed to match for identical systems (maybe just a json transcription of lstopo or something) If all ranks write their blob then do a kvs_fence we might have a pretty scalable resource gather operation.

That should work, but if the module is loaded from the "init script", some additional synchronization will be needed since `flux module load` will return before the fence completes.  The script could watch for the KVS resource directory to become non-NULL and then by virtue of all the updates happening under one fence, the entire hierarchy should be complete.

Probably the module should only do the fence if a count argument is passed in at module load time, otherwise reloading the module on a subset of ranks would hang.  Just do a commit otherwise.

**Comment by garlick** [comment]
The user can also launch multple ranks per node via Slurm, so it almost seems like each broker should try to figure out what its slurm allocation is.

**Comment by trws** [comment]
If run with `--size`, maybe we should just say that each gets one logical core?  When running via slurm, @garlick makes a good point, we could pull the allocation out of the slurm environment.  That might really simplify things.  Given those base conditions, the rest of the information should be easy to get from hwloc if we're willing to pull that in as a dependency, or a subset of it could be pulled in from libnuma and then enhanced with cpuid to identify logical/physical cores.

**Comment by grondo** [comment]
SLURM also sets cpu_affinity by default... tasks don't get a real
"allocation" in SLURM afaik. On clusters using consumable resources plugin,
the cpuset and memory cgroups may reflect that task's allocated resources

**Comment by grondo** [comment]
> If run with --size, maybe we should just say that each gets one logical
> core?
>
How would each broker decide *which* core, and for NUMA, on which socket?

**Comment by trws** [comment]
Since a user can run an arbitrarily high number, I wasn't really 
thinking about that issue.  More that it seemed sensible for each broker 
to say that its resource pool had one core, since you can't really drop 
below one usefully.

**Comment by grondo** [comment]
> Since a user can run an arbitrarily high number, I wasn't really
> thinking about that issue. More that it seemed sensible for each broker
> to say that its resource pool had one core, since you can't really drop
> below one usefully.


Oh, yeah. That is reasonable. I was thinking in terms of the resource
representation.
It might also be nice to be able to "fake" the layout/cores with
`flux-start --size` for testing
purposes.

**Comment by trws** [comment]
Very good point.

As to actual implementation, I would probably want the ranks to do a round-robin across resources in a node.  More specifically, something equivalent to the `spread` affinity policy from the OpenMP spec, meaning something like "round-robin across resources placing each subsequent resource farthest in the cache hierarchy as possible from the previous."  On a two socket eight core system with caches shared by core pairs and a direct map of cores to caches to sockets in order four four processes would be mapped to 0, 4, 2, 6. This is one place where hwloc would be nice, it implements a number of policies like this internally such that we wouldn't have to redo all of that.

**Comment by garlick** [comment]
A "resource" service is proposed in #296.  Maybe this thing should just be called "hwloc" not "resrc"?

**Comment by trws** [comment]
If we're planning to do overridable services, perhaps it should be the `resource_hwloc` module, providing the `resource` service?

**Comment by grondo** [comment]
> If we're planning to do overridable services, perhaps it should be the resource_hwloc module, providing the resource service?

@trws, yeah that is what I was thinking.

First thing is to get the data into the instance. @garlick might be right that this could be stored under `hwloc` tree in the kvs, but really the name doesn't matter that much to me. Then a module `resource-hwloc` could be loaded which would provide the `resource` service using the data from `hwloc` location in the kvs. (or it could all be one service)

For the short term milestone, however, I wonder if overridable services will be ready. If not, should the contingency be to have the scheduler read from resource-hwloc directly?

**Comment by trws** [comment]
I would kinda rather that we use the resource protocol, and just say that for now only complete overriding of a service is supported.  That means that the RDL module wouldn't be able to use the resource interface to talk to hwloc, but it could read from the KVS if it so desires, or use another side channel.  It wouldn't be inheritance, but it would still give us the benefits of a unified protocol.  Perhaps more like a Go interface?

**Comment by grondo** [comment]
Sounds good.

**Comment by trws** [comment]
closed by #312
